src/crypto/isa-l/md5/async_md5.cc

```diff
@@ -0,0 +1,160 @@
+ * the Free Software Foundation.  See file COPYING.
+ *
+ */
+
+#include "async_md5.h"
+
+#include <concepts>
+#include <span>
+#include <boost/asio/append.hpp>
+#include <boost/asio/dispatch.hpp>
+
+namespace ceph::async_md5 {
+
+boost::system::error_category& hash_category()
+{
+  static struct category : boost::system::error_category {
+    const char* name() const noexcept override { return "isal-hash"; }
+    std::string message(int code) const override {
+      switch (code) {
+        case HASH_CTX_ERROR_NONE: return "none";
+        case HASH_CTX_ERROR_INVALID_FLAGS: return "invalid flags";
+        case HASH_CTX_ERROR_ALREADY_PROCESSING: return "already processing";
+        case HASH_CTX_ERROR_ALREADY_COMPLETED: return "already completed";
+        default: return "unknown";
+      }
+    }
+  } instance;
+  return instance;
+}
+
+
+Digest::Digest()
+{
+  hash_ctx_init(&ctx);
+  ctx.user_data = this;
+}
+
+Digest& digest_from_hash_ctx(MD5_HASH_CTX* ctx)
+{
+  return *reinterpret_cast<Digest*>(ctx->user_data);
+}
+
+// copy the given byte as two hex characters to output
+template <std::output_iterator<char> Iterator>
+Iterator hex_encode(Iterator out, unsigned char b)
+{
+  static constexpr char hexdigits[16] = {
+    '0', '1', '2', '3', '4', '5', '6', '7',
+    '8', '9', 'a', 'b', 'c', 'd', 'e', 'f'
+  };
+  *out++ = hexdigits[b >> 4]; // high 4 bits
+  *out++ = hexdigits[b & 0xf]; // low 4 bits
+  return out;
+}
+
+static std::string to_hex(std::span<const MD5_WORD_T> digest)
+{
+  auto result = std::string(32, '\0');
+  std::string::iterator out = result.begin();
+
+  for (const MD5_WORD_T w : digest) {
+#if defined(__BYTE_ORDER__) && defined(__ORDER_BIG_ENDIAN__) \
+    && __BYTE_ORDER__ == __ORDER_BIG_ENDIAN__
+    out = hex_encode(out, w >> 24);
+    out = hex_encode(out, w >> 16);
+    out = hex_encode(out, w >> 8);
+    out = hex_encode(out, w >> 0);
+#else
+    out = hex_encode(out, w >> 0);
+    out = hex_encode(out, w >> 8);
+    out = hex_encode(out, w >> 16);
+    out = hex_encode(out, w >> 24);
+#endif
+  }
+  return result;
+}
+
+std::string Digest::as_hex() const
+{
+  return to_hex(hash_ctx_digest(&ctx));
+}
+
+
+Batch::Batch(const executor_type& ex, std::chrono::nanoseconds batch_timeout)
+    : ex(ex), timer(ex), batch_timeout(batch_timeout)
+{
+  md5_ctx_mgr_init(&mgr);
+}
+
+void Batch::complete(MD5_HASH_CTX* ctx, MD5_HASH_CTX* submit_ctx)
+{
+  error_code ec;
+  if (int err = hash_ctx_error(ctx);
+      err != HASH_CTX_ERROR_NONE) {
+    ec.assign(err, hash_category());
+  }
+  auto& digest = digest_from_hash_ctx(ctx);
+  auto h = boost::asio::append(std::move(digest.handler), ec);
+
+  if (ctx == submit_ctx) {
+    // if we're completing the ctx inside of its own call to async_hash(), we
+    // have to post() the completion instead
+    boost::asio::post(boost::asio::bind_executor(get_executor(), std::move(h)));
+  } else {
+    boost::asio::dispatch(std::move(h));
```

Comment:
while testing https://github.com/ceph/ceph/pull/52488, we sometimes see a crash here due to what looks like an empty any_completion_handler

in the trace i'm looking at, `Batch::init_async_hash()` was calling `complete()` on the ctx returned by `md5_ctx_mgr_submit()`, which was not the same as `submit_ctx`

Reply:
i added a unit test that can reproduce this crash. after enabling the thread sanitizer:
```diff
diff --git a/src/test/isal/CMakeLists.txt b/src/test/isal/CMakeLists.txt
index bbff3bd0d03..1b83b89d14a 100644
--- a/src/test/isal/CMakeLists.txt
+++ b/src/test/isal/CMakeLists.txt
@@ -1,3 +1,5 @@
 add_executable(unittest_isal_md5 test_isal_md5.cc)
 add_ceph_unittest(unittest_isal_md5)
 target_link_libraries(unittest_isal_md5 isal_md5)
+target_compile_options(unittest_isal_md5 PRIVATE "-fsanitize=thread")
+target_link_options(unittest_isal_md5 PRIVATE "-fsanitize=thread")
```
and running that test case repeatedly:
```
bin/unittest_isal_md5 --gtest_filter=isal_md5.batch_thread_pool --gtest_repeat=100 &> batch_thread_pool.log
```
thread sanitizer reports these data races: https://gist.github.com/cbodley/5f4dcc6b1148b2302313a5ba7edc1c15